### ISSUE TITLE
Learning Data Augmentation Strategies for Object Detection？

### ISSUE THREAD

**Issue by feixiangdekaka** [tensorflow/tpu#423]
Where is the code for implement the 《Learning Data Augmentation Strategies for Object Detection》?

**Comment by saberkun** [comment]
When using augmentation, it triggers
https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py#L15

Yaml config is https://github.com/tensorflow/tpu/blob/master/models/official/detection/configs/yaml/retinanet_autoaugment.yaml

**Comment by feixiangdekaka** [comment]
> When using augmentation, it triggers
> https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py#L15
> 
> Yaml config is https://github.com/tensorflow/tpu/blob/master/models/official/detection/configs/yaml/retinanet_autoaugment.yaml
in the function:
def distort_image_with_autoaugment(image, bboxes, augmentation_name):
  """Applies the AutoAugment policy to `image` and `bboxes`.
  Args:
    image: `Tensor` of shape [height, width, 3] representing an image.
    bboxes: `Tensor` of shape [N, 4] representing ground truth boxes that are
      normalized between [0, 1].

bboxes is the format of [x1,y1,x2,y2]  ?  or other format?

**Comment by gittigxuy** [comment]
So,how to use this autoaugment_util.py  file?could you please give me a example?Thanks

**Comment by feixiangdekaka** [comment]
some result like this  ,something is right,others are not,.
![3](https://user-images.githubusercontent.com/45898856/60398029-51b12e80-9b86-11e9-971c-8f6c49ab6268.jpg)
![0](https://user-images.githubusercontent.com/45898856/60398037-5d9cf080-9b86-11e9-954f-d6b5c729e4f8.jpg)
![9](https://user-images.githubusercontent.com/45898856/60398039-642b6800-9b86-11e9-9c88-408176852df2.jpg)

**Comment by heartInsert** [comment]
> So,how to use this autoaugment_util.py file?could you please give me a example?Thanks

In   detection.dataloader.retinanet_parser.py    line  221

`  if self._use_autoaugment:

      image, boxes = autoaugment_utils.distort_image_with_autoaugment(
          image, boxes, self._autoaugment_policy_name)`

make     self._autoaugment_policy_name='v0'

**Comment by poodarchu** [comment]
does it work on coco?
I found several bugs in autoaugment_utils.py.

**Comment by BarretZoph** [comment]
Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?

**Comment by heartInsert** [comment]
I  think  I  didn"t   meet   bugs   here

**Comment by poodarchu** [comment]
> Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?

1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.

**Comment by heartInsert** [comment]
> > Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?
> 
> 1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
> 2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.

I   can't  reply   you  the  first  question,.
The  second  question's  answer is  that :    it's   not  wrong ,   this   is   what    augment    should   do .This  situation  has   been  displayed   in   their   paper.

**Comment by poodarchu** [comment]
> > > Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?
> > 
> > 
> > 
> > 1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
> > 2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.
> 
> I can't reply you the first question,.
> The second question's answer is that : it's not wrong , this is what augment should do .This situation has been displayed in their paper.

But in my experiments, If I do not fix these 'bugs', mAP on COCO only achieves 10.5%。after I fix these, mAP start to be resonable.

**Comment by heartInsert** [comment]
> > > > Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?
> > > 
> > > 
> > > 
> > > 1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
> > > 2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.
> > 
> > 
> > I can't reply you the first question,.
> > The second question's answer is that : it's not wrong , this is what augment should do .This situation has been displayed in their paper.
> 
> But in my experiments, If I do not fix these 'bugs', mAP on COCO only achieves 10.5%。after I fix these, mAP start to be resonable.

When  these  two  bugs  which  you  said above   been   applied  in  training  ,   it   means  the  unstable   increased  in  training  data  .    Did  you   run   more   5 or 10  epoch   in  training  . If   MAP   in  test_data    correspoinding    increased  ,    it   meas   it's  not   a   bug  , or if  MAP  doesn't   changed ,I  think  you  are  right .
If  you  do  these   experiment  above , please inform  me.

**Comment by poodarchu** [comment]
> > > > > Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?
> > > > 
> > > > 
> > > > 
> > > > 1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
> > > > 2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.
> > > 
> > > 
> > > I can't reply you the first question,.
> > > The second question's answer is that : it's not wrong , this is what augment should do .This situation has been displayed in their paper.
> > 
> > 
> > But in my experiments, If I do not fix these 'bugs', mAP on COCO only achieves 10.5%。after I fix these, mAP start to be resonable.
> 
> When these two bugs which you said above been applied in training , it means the unstable increased in training data . Did you run more 5 or 10 epoch in training . If MAP in test_data correspoinding increased , it meas it's not a bug , or if MAP doesn't changed ,I think you are right .
> If you do these experiment above , please inform me.

Have you tried on COCO?

**Comment by heartInsert** [comment]
> > > > > > Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?
> > > > > 
> > > > > 
> > > > > 
> > > > > 1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
> > > > > 2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.
> > > > 
> > > > 
> > > > I can't reply you the first question,.
> > > > The second question's answer is that : it's not wrong , this is what augment should do .This situation has been displayed in their paper.
> > > 
> > > 
> > > But in my experiments, If I do not fix these 'bugs', mAP on COCO only achieves 10.5%。after I fix these, mAP start to be resonable.
> > 
> > 
> > When these two bugs which you said above been applied in training , it means the unstable increased in training data . Did you run more 5 or 10 epoch in training . If MAP in test_data correspoinding increased , it meas it's not a bug , or if MAP doesn't changed ,I think you are right .
> > If you do these experiment above , please inform me.
> 
> Have you tried on COCO?

COCO  is   a  little  large   for  my  1050ti  （：

**Comment by poodarchu** [comment]
I've convert the tf version to numpy, so everyone can run using any dl framework.

https://github.com/poodarchu/learn_aug_for_object_detection.numpy

**Comment by heartInsert** [comment]
Great  work ,thanks  for  sharing

**Comment by sefira** [comment]
https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py#L686

I think I've found another bug for Learning Data Augmentation Strategies for Object Detection seen above link, in which an image is augmented as many time as the number of bboxes in it.
Assuming we are augmenting the bboxes and images by random flipping, in this case, each bbox is flipped by a probability P, but the image will be flip many times and the probability is not equal to P.

**Comment by BarretZoph** [comment]
> > Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?
> 
> 1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
> 2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.

Yes these two things do occur, but the performance is still good on COCO. Some of these could actually potentially benefit the training by acting as a regularizer. 

I have since fixed 2 and noticed that on COCO the performance on ResNet-50 improves by about 0 to 0.2 mAP depending on the augmentation policy. I am playing around with both fixes to see if the performance could be better improved still, but the performance in the paper is obtained with both of these features in it.

**Comment by BarretZoph** [comment]
> https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py#L686
> 
> @BarretZoph @aman2930 @saberkun I think I've found another bug for Learning Data Augmentation Strategies for Object Detection seen above link, in which an image is augmented as many time as the number of bboxes in it.
> Assuming we are augmenting the bboxes and images by random flipping, in this case, each bbox is flipped by a probability P, but the image will be flip many times and the probability is not equal to P.

I am not sure I understand this. There operations that are applied I.I.D to each bbox with the image.

**Comment by Jacobew** [comment]
Hi, is there any pytorch implementation of [https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py](url)? 
If not, I'll appreciate if anyone could give me some advice on converting this tf code into pytorch version. :)

**Comment by Jacobew** [comment]
> > > > Yes it does work on COCO. What bugs did you find in autoaugment_utils.py?
> > > 
> > > 
> > > 
> > > 1. When it does bbox only flip / shear / translate, if a bbox has box in it, it will cause the smaller bboxes mismatch with correspoinding objects.
> > > 2. if a bbox is out of image range after a transform, the box will still exist in image, and it's shape is totally wrong.
> > 
> > 
> > I can't reply you the first question,.
> > The second question's answer is that : it's not wrong , this is what augment should do .This situation has been displayed in their paper.
> 
> But in my experiments, If I do not fix these 'bugs', mAP on COCO only achieves 10.5%。after I fix these, mAP start to be resonable.

@poodarchu What AP did you get when these bugs got fixed? BTW, could you share the fixed version? Thanks.

**Comment by BarretZoph** [comment]
Just to clarify, with the two issues listed above, you should be able to reproduce the results in the original paper. When fixing them I found little to no mAP performance improvement on COCO.

**Comment by Qianshaowei** [comment]
> When using augmentation, it triggers
> https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py#L15
> 
> Yaml config is https://github.com/tensorflow/tpu/blob/master/models/official/detection/configs/yaml/retinanet_autoaugment.yaml

How to learn Data Augmentation Strategies

> When using augmentation, it triggers
> https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py#L15
> 
> Yaml config is https://github.com/tensorflow/tpu/blob/master/models/official/detection/configs/yaml/retinanet_autoaugment.yaml

Excuse me, where is the code of the RNN controller?

**Comment by Qianshaowei** [comment]
> When using augmentation, it triggers
> https://github.com/tensorflow/tpu/blob/master/models/official/detection/utils/autoaugment_utils.py#L15
> 
> Yaml config is https://github.com/tensorflow/tpu/blob/master/models/official/detection/configs/yaml/retinanet_autoaugment.yaml

Excuse me, where is the code of the RNN controller?

**Comment by noreenanwar** [comment]
This method works for small objects?
